File: binarysearch/34-find-first-and-last-position/FindFirstAndLastPosition.py
class Solution:
    def searchRange(self, nums: List[int], target: int) -> List[int]:
        def findBounds(isFirst: bool)->int:
            left, right = 0, len(nums)-1 
            bound = -1
            while left<=right:
                mid = (left+right)//2
                if nums[mid]==target:
                    bound = mid
                    if isFirst:
                        right = mid - 1
                    else:
                        left  = mid+1
                elif nums[mid]>target:
                    right = mid -1
                else:
                    left = mid + 1
            return bound
        return [findBounds(True), findBounds(False)]

File: binarysearch/34-find-first-and-last-position/FindFirstAndLastPosition.java
class Solution {
    public int[] searchRange(int[] nums, int target) {
        int first = findbounds(nums, target, true);
        int last = findbounds(nums, target, false);
        return new int[]{first, last};
    }
    private int findbounds(int nums[], int target, boolean isFirst){
        int left = 0, right = nums.length-1, bound=-1;
        while(left<=right){
            int mid = left +(right - left)/2;
            if(nums[mid]==target){
                bound = mid;
                if(isFirst){
                    right  = mid-1;
                }else{
                    left = mid +1;
                }
            }else if(nums[mid]>target){
                right = mid-1;
            }else{
                left = mid+1;
            }
        }
        return bound;
    }

}